### ISSUE TITLE
Global handling of phase-encode direction data

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#747]
Following on from #591, and something that has come up during the process of #725.

Providing an automated processing pipeline where you don't necessarily know _a priori_ what type of reversed phase-encode acquisition has taken place, or what direction(s) the phase encoding directions are in, or whether a fieldmap has instead been acquired. In such cases the current `dwipreproc` interface is limited.

I've started on code to extract the phase encode direction and readout bandwidth information from DICOMs; and this information is expected to be provided in BIDS compliant data. But given the number of possible combinations of such data, it may be preferable to instead embed this information within image headers, so that `dwipreproc` can figure out what it has to do (this would be in addition to the current 'manual' options for specifying phase-encoding information).

One option would be to embrace the approach in #591, adopting a wider 'standard' of having phase encoding information embedded within the '`dw_scheme`' header entry. Alternatively, particularly since the DWI scheme is now a plain key/value entry in the header rather than a dedicated field, we could add a '`pe_scheme`' entry. I think I'm in favour of the latter; not only for separating the two sources of information, but also the DW scheme is specified with respect to scanner space whereas the PE scheme is with respect to image axes, so combining them is a bit of a hack.

There's a semi-standardised header field '`"PhaseEncodingDirection`' (`i`, `i-`, `j`, `j-`, `k`, `k-`) that may be specified for any particular image file, but obviously this can't vary for different volumes in a 4D image. I could _technically_ get away with just using this information, but it would require a _lot_ of gymnastics in order to figure out how to invoke `dwipreproc`, and wouldn't be guaranteed to work for all cases. Plus, given whatever evil scheme @jdtournier has going that spawned #591 requires a PE direction per volume, I reckon it looks like a good idea to support that longer-term.

Opening up for discussion / comments. If I go for my preferred option ('`PhaseEncodingDirection`' for fixed PE direction, '`pe_scheme`' for variable), @jdtournier you'll need to warp your code around that accordingly.

**Comment by Lestropie** [comment]
Notes:

* Would revert #591 and enforce 4 columns in the `dw_scheme` entry if going for the `pe_scheme` option.

* Commands such as `dwiextract`, `mrconvert`, `mrcat` would need to be aware of the `PhaseEncodingDirection` / `pe_scheme` header entries and behave accordingly.

* `dwiextract` could select and extract volumes based on phase encoding direction.

* Propagating / deleting these fields by certain operations would probably mimic the result of #375.

**Comment by jdtournier** [comment]
Happy with most of that, just a couple of comments:

> There's a semi-standardised header field '"PhaseEncodingDirection' (i, i-, j, j-, k, k-) that may be specified for any particular image file, but obviously this can't vary for different volumes in a 4D image.

I don't get that. If it's a DICOM entry, it will be specified per slice, so there's nothing preventing it from varying volume to volume even within a single DICOM series - although currently I don't anticipate any vendors would provide a standard sequence capable of doing this in their product. But it can be implemented (as we have done), and would therefore be the correct place to look for this information. 

> we could add a 'pe_scheme' entry.

I agree this would be the most flexible. Note that it'll need a little bit more gymnastics to get right. I have a feeling the best thing to do would be to have each line in the `pe_scheme` provide the PE direction scaled by the bandwidth - as required by EDDY for instance. 

Only question is whether we adopt a TOPUP `acqp` type approach, providing a list of the actual PE directions & bandwidths used upfront, together with a per-volume index file. Personally, I think it would be more robust to just provide a full per-volume PE entry, but just thought I'd throw that out there. 

> I could technically get away with just using this information, but it would require a lot of gymnastics in order to figure out how to invoke dwipreproc

Not sure I completely understand why this would be a problem. If we store the information within the headers, then `dwipreproc` should be able to accept a single 4D dataset with mixed PE directions, and feed that through to TOPUP/EDDY as appropriate? 

> whatever evil scheme @jdtournier has going that spawned #591 

Bit harsh... 

> Would revert #591 and enforce 4 columns in the dw_scheme entry if going for the pe_scheme option.

I wouldn't. There's no ambiguity in leaving remaining columns as-is (they're just unused by the standard _MRtrix3_ core), it requires no modifications to the code (well, I did have to relax some of the checks a touch, but it wasn't exactly a big deal), is already implemented, and provides flexibility for further uses that might come up. I certainly found it very useful to handle our data...

**Comment by Lestropie** [comment]
> I don't get that. If it's a DICOM entry, it will be specified per slice, so there's nothing preventing it from varying volume to volume even within a single DICOM series - although currently I don't anticipate any vendors would provide a standard sequence capable of doing this in their product. But it can be implemented (as we have done), and would therefore be the correct place to look for this information.

By 'semi-standardised', I mean that it's in the BIDS standard, and it looks like other DICOM conversion tools are updating their outputs based on those requirements. If our DICOM importer outputs this information in any other field, then any downstream operation that depends on this information will have to look in multiple fields. So it makes sense to me that if the phase encode direction remains fixed throughout a series, this would be the place to store that.

The fact that it's specified per slice in the DICOMs doesn't really matter (there's plenty of other DICOM information that our importer just grabs from the first slice and ignores the rest). Unless you actually want to capture per-slice variations in that parameter from the image headers, which is over and above storing the phase encode direction per volume in the image header under discussion here, which itself is above and beyond what others are doing (storing a single `PhaseEncodeDirection` for the entire volume series).

> I agree this would be the most flexible. Note that it'll need a little bit more gymnastics to get right. I have a feeling the best thing to do would be to have each line in the pe_scheme provide the PE direction scaled by the bandwidth - as required by EDDY for instance.

I was thinking of just using the `eddy` format, since that's the primary tool currently using that info: 3 columns for unit vector PE direction, then a total readout time. Technically we could do differently, but there'd need to be a justification for it.

> Only question is whether we adopt a TOPUP acqp type approach, providing a list of the actual PE directions & bandwidths used upfront, together with a per-volume index file. Personally, I think it would be more robust to just provide a full per-volume PE entry, but just thought I'd throw that out there.

I concur with per-volume entries. Easy to convert to a reduced list and index file if necessary, cleaner for e.g. adding / extracting volumes, and consistent with `dw_scheme`.

> > I could technically get away with just using this information, but it would require a lot of gymnastics in order to figure out how to invoke dwipreproc

> Not sure I completely understand why this would be a problem. If we store the information within the headers, then dwipreproc should be able to accept a single 4D dataset with mixed PE directions, and feed that through to TOPUP/EDDY as appropriate?

As in, if _only_ `PhaseEncodeDirection` can be used (which must by design be fixed for the entire volume series), and `pe_scheme` or similar doesn't exist, then it becomes clumsy trying to invoke `dwipreproc` when you don't know _a priori_ what type of phase-encoding acquisition design has been used. If you _have_ `pe_scheme`, then you can just concatenate everything together and send to `dwipreproc`, and it'll do the heavy lifting.

> > whatever evil scheme @jdtournier has going that spawned #591 

> Bit harsh...

That was a compliment. Evil schemes are awesome.

> > Would revert #591 and enforce 4 columns in the dw_scheme entry if going for the pe_scheme option.

> I wouldn't. There's no ambiguity in leaving remaining columns as-is (they're just unused by the standard MRtrix3 core), it requires no modifications to the code (well, I did have to relax some of the checks a touch, but it wasn't exactly a big deal), is already implemented, and provides flexibility for further uses that might come up. I certainly found it very useful to handle our data...

I suppose the question is: If the use for additional columns in `dw_scheme` is _not_ phase-encoding information (which would be clumsy to continue storing in `dw_scheme` once `pe_scheme` is available), is it something that remains _appropriate to continue storing_ in `dw_scheme` (and therefore retain the relaxed column count), or is that information something that should additionally be moved to its own header entry (and then revert `dw_scheme` to be forced to 4 columns in order to detect possible conversion / formatting errors)?

Only thing that comes to mind is non-PGSE diffusion, in which case you might store OGSE frequency or second diffusion sensitisation direction & _b_-value or the like. If so, it might need some more consideration: e.g. throw a warning if doing some operation on the presumption of a PGSE sequence but `dw_scheme` contains more than 4 columns.

**Comment by jdtournier** [comment]
> By 'semi-standardised', I mean that it's in the BIDS standard,

Ah, sorry, I'd assumed you mean that the `PhaseEncodeDirection` was a DICOM entry - my mistake. Makes sense now. 

> I was thinking of just using the eddy format, since that's the primary tool currently using that info: 3 columns for unit vector PE direction, then a total readout time.

Once again, my mistake - I'd forgotten that EDDY requires a 4 column table. Fully agree with that format.

> f only PhaseEncodeDirection can be used (which must by design be fixed for the entire volume series), and pe_scheme or similar doesn't exist, then it becomes clumsy trying to invoke dwipreproc when you don't know a priori what type of phase-encoding acquisition design has been used.

Yep, I get that - based on my earlier misunderstanding...

> Evil schemes are awesome.

You voting Trump then...? :wink:

> If the use for additional columns in dw_scheme is not phase-encoding information (which would be clumsy to continue storing in dw_scheme once pe_scheme is available), is it something that remains appropriate to continue storing in dw_scheme (and therefore retain the relaxed column count), or is that information something that should additionally be moved to its own header entry (and then revert dw_scheme to be forced to 4 columns in order to detect possible conversion / formatting errors)?

I totally agree that if we go with the `pe_scheme` entry, there'd be no need to maintain that information in the `dw_scheme` entries. What I'm arguing is that there's no harm in allowing additional columns in the `dw_scheme` table for whatever purpose people might come up with. The PE direction thing is actually a good example of using that facility to quickly store custom per-volume data for a specific project, which can then be formalised into their own dedicated entries if and when we decide it would be valuable to do so in general. Maybe someone will come up with a fancy sequence with different echo times, diffusion times, or whatever, and the `dw_scheme` table would provide a really simple means of storing that information for their project. These additional columns don't interfere with normal operation, so there's no reason to actively remove them.

**Comment by Lestropie** [comment]
> Ah, sorry, I'd assumed you mean that the PhaseEncodeDirection was a DICOM entry - my mistake. Makes sense now. 

`0018 1312` is DICOM for 'In-plane Phase Encoding Direction', and comes in either '`ROW`' or '`COL`' flavour - which isn't of much use outside of a DICOM. Siemens have a CSA entry '`PhaseEncodingDirectionPositive`' to give the polarity - haven't actually tested it though.

I don't know for sure if '`PhaseEncodingDirection`' originates from the process of defining the BIDS standard, or elsewhere beforehand.

> These additional columns don't interfere with normal operation, so there's no reason to actively remove them. 

Main thing that came to mind is if people are trying to manually construct a gradient table from whatever, and are buggering it up. Or maybe erroneously transposing their data _a la_ [this post](http://community.mrtrix.org/t/error-mrconvert-cannot-import-diffusion-gradient-table/363/3): they'd currently get 'number of diffusion directions doesn't match number of volumes', which may or may not be enough of a hint. If you don't think that's enough of a concern to reinstate the restriction, that's fine, I'm happy to run as-is; I'll even design the `pe_scheme` handling correspondingly.

I suppose I thought one advantage of `dw_scheme` being just a header keyval entry is that you can use the same technique to store *any* per-volume parameters; you're not 'locked' to using the DW scheme just because it's the only available dynamically-sized matrix entry. But we'd need to modify `dwiextract` / `mrcat` / `mrconvert` to detect and handle such _a-priori_-unknown fields.

**Comment by thijsdhollander** [comment]
Slightly veering off topic, but on this one:

> "number of diffusion directions doesn't match number of volumes"

...it may actually appear slightly cryptic in various scenarios.  A helpful addition may be that the command also states how many directions actually were detected, as well as the number of volumes.  (in a script I've recently written, I actually added this kind of feedback at the start)

This is a bit in line with the recent addition that commands that error out based on the number of provided arguments being wrong, also show what those commands interpreted as being the provided arguments.

**Comment by Lestropie** [comment]
I've been chipping away at [the new version of `dwipreproc`](https://github.com/MRtrix3/mrtrix3/blob/stanford/scripts/dwipreproc) for a while, it's now at the point where it can actually be used and has played ball with the data I have thrown at it thus far. Apart from requesting some help RE non-Siemens data, I also wanted to explain both the design around managing phase encoding schemes and the corresponding changes to `dwipreproc` as they stand, see if anybody has opinions or suggestions.

#### Tracking phase encoding information

With the [BIDS standard](http://bids.neuroimaging.io/), they introduced two new 'standardized' key/value fields: `PhaseEncodingDirection` and `TotalReadoutTime`. Other softwares are already 'supporting' them to some extent or another. This is enough information to perform susceptibility distortion correction based on header information only (well, in BIDS they're stored in a JSON file associated with the NIfTI, but for basic capabilities that's basically equivalent to placing key/value pairs in an external file rather than embedding in the header), without requiring that the user (or script) explicitly specify the type of acquisition protocol that has taken place.

However: This only allows one PE direction / readout time configuration per image file. This is bad for more advanced applications, as you may want volumes with different phase encoding parameters within the one file (or may even acquire them as such). It was also going to make writing my BIDS pipeline a nightmare, as there's an infinite number of ways that images with different phase encoding could be acquired in a protocol and fed to the pipeline, and the pipeline doesn't know _a priori_ what it's going to get. It would also be preferable, if we get around to trying our own image pre-processing, to be able to feed in a single image file, and have all of the phase-encoding information there in the header ready to go.

So what I've done is hopefully a compromise that will enable all of the things: make my BIDS pipeline script actually feasible, make use of the data fields expected to be provided in JSON files, make data exported from MRtrix compatible with other software packages if they are updated to support BIDS, but also enable different phase encoding parameters for different volumes within a file.

* A phase encoding 'scheme' matrix is defined, much like the diffusion weighting scheme is currently. It can be three columns (direction only; needs testing), four columns (direction & readout time), or more columns (for reasons unknown).

* For DICOM import, the phase encoding scheme is determined as best as possible using the DICOM fields. This will require testing, and anyone using a non-Siemens scanner would need to determine the appropriate fields to capture for the data they're acquiring.

* Wrapper functions are defined for handling any phase encoding information present in the header (or imported into a header from a JSON file before the fact) in new namespace [`MR::PhaseEncoding`](https://github.com/MRtrix3/mrtrix3/blob/stanford/lib/phase_encoding.h):

  * Phase encoding schemes are _always_ handled using a matrix (in order to deal with the most complex use case); it's only the interaction with the header key/value entries that changes depending on the details. When a developer requests the phase encoding scheme, they get the full matrix, regardless of what is stored in the header. When the scheme is stored, the wrapper function receives a matrix and determines what should go into the header.

  * If the phase encoding is equivalent for all volumes, then fields `PhaseEncodingDirection` and `TotalReadoutTime` are stored in the header keyval list.

  * If the phase encoding _varies_ between volumes, then field `pe_scheme` stores the complete matrix (much like `dw_scheme` does currently; note however that `pe_scheme` directions are defined with respect to RAS axes).

* Any function that may deal with addition or removal of volumes (e.g. `mrconvert`, `mrcat`) will attempt to preserve the phase encoding scheme, just as is currently done with the diffusion gradient scheme.

* The phase encoding scheme can additionally be imported / exported from / to `eddy`'s 'config' / 'index' format using explicit command-line options.

#### `dwipreproc` rewrite

With my BIDS pipeline, I needed the ability to simply pass the DWI data to `dwipreproc`, with the possibility of specifying additional image data purely for estimating the inhomogeneity field. So this requires examining the phase encoding scheme present, choosing what data to use in estimating the inhomogeneity field (if any), converting the phase encoding scheme & image data into the formats expected by `topup` / `applytopup` / `eddy`, and choosing whether or not explicit recombination of volumes with equivalent diffusion direction but opposing phase encode direction should be performed.

By the time I went to the effort of doing this, it seemed unusual to keep the existing script paths for the `-rpe_none` / `-rpe_pair` - `rpe_all` options, since there was actually not that much overlap between them as it was, and the command-line interface wasn't really compatible with the new approach of extracting this information from the header. Moreover, there's a possibility of explicitly activating one of these options, and that configuration _conflicting_ with what's stored in the images header, which warrants a user warning.

Therefore, the new approach to these script options is to _construct the phase-encoding table(s) based on user input_. Once this is done, the script can subsequently use essentially the same code as that used when this information is available in the header to begin with.

There are a few little caveats though, particularly with interface, which I know @thijsdhollander won't like :-P But hey, it's a script provided on good faith for convenience, and I'm hoping that over time more data will be converted from DICOM with this information in the header, so the existing options will be used less over time.

* New option `-rpe_header` indicates that any phase encoding information is embedded within the image header.

* The phase encoding direction is no longer a compulsory argument. Option `-pe_dir` becomes compulsory whenever `-rpe_none` / `-rpe_pair` / `-rpe_all` are used. `-rpe_header` and `-pe_dir` are mutually exclusive options (since `-rpe_header` expects phase encoding information to already be present in the header).

* The total readout time can be provided using the `-readout_time` option. If it's not provided, the previous 'sane' default of 0.1 will be used and a warning of such will be issued. This also cannot be specified in conjunction with `-rpe_header`.

* Option `-topup_images` is used to provide an image series that is to be used for field estimation, but will _not_ form part of the output DWIs. If `-topup_images` is _not_ provided, the script will extract the _b_=0 volumes from the input DWIs and see if there's any phase encoding contrast to work with. If not, user will receive a warning that no field estimation is taking place.

* `-rpe_pair` previously expected a pair of image files following the option; one for _b_=0 with the same phase encoding as the input DWIs, one for _b_=0 with the opposite phase encoding. These images should now instead be concatenated and provided via `-topup_images`.

* `-rpe_all` previously expected one image file following the option, being a second DWI series with exactly the same diffusion directions as the input DWI but opposing phase encoding. These data should now instead be concatenated and provided as a single input DWI file to the script.

With the last two, although _theoretically_ I could have kept the same interface as before for consistency, I don't think it would work out quite as well. E.g. Having `-rpe_pair` expect two image inputs despite there being a `-topup_images` option available would be weird. I never liked how `-rpe_all` only provided half of the DWI data as the 'input' and then the second half as 'optional input'. Also this way, `-rpe_none` / `-rpe_pair` / `-rpe_all` / `-rpe_header` are all _option flags_, which is nice and consistent. In fact setting the phase encoding acquisition design could now plausibly become a compulsory positional argument rather than four options...

**Comment by thijsdhollander** [comment]
Just to make everyone sleep well during at least the next night in whatever time zone everyone lives in: I'm generally ok with these interface changes. :wink:

What I'm mostly sensitive to (yes, I'm a sensitive guy :blush:), is changes that force a certain logic upon the user, or that force them to deal with a logic (i.e. be confronted with an feel the need to understand before proceeding) that was not needed at all for a common more simple scenario.  In this case, the "simple scenarios" would probably be the 3 typical ones we offered before: a majority of users still find themselves in one of these 3, and the existing command was actually a blessing in the sense that it made these as simple as I can image ever getting them.  Definitely much better than going through the sheer hell of doing all the manual interactions with topup/eddy themselves.

The changes that you propose still have these options (even though the actual interface is not *exactly* backwards compatible).  People can still do what they did before, but the names of the options introduced also *don't make this feel like they are dealing with a different logic* (compared to before).

I haven't gone through the whole post, but from what I see, the previous use of something along the lines of
```
dwipreproc AP series-ap.mif corrected.mif -rpe_pair ap.mif pa.mif
```
...becomes...
```
dwipreproc series-ap.mif corrected.mif -rpe_pair -topup_images ap-pa.mif
```
...right?

That definitely feels ok.

A few random comments (but I may have not picked up on the answers maybe inherently being in the post):
- I would change the name of the option `topup_images` though.  Basically to avoid a reference to the specific software, and rather describe these actual images.  Maybe `rpe_images` or something... lacking further inspiration at the moment.
- Is `topup_images` actually used for anything else than in the `rpe_pair` context?
- Before (i.e., take my initial example above for what follows), one could supply multiple images to `rpe_pair`, e.g. `ap.mif` and `pa.mif` could have more than one (b=0) image each, using the 4th dimension.  How would this work via the `topup_images` option, taking just one image?  Let's say I had 3 b=0 images for each (of 2) phase encoding direction.  Do I simply concatenate them, offering a volume with 6 images in the 4th dimension to `topup_images`?

And maybe a more general topic to think about: what do we actually advise users to do as a denoising strategy on all these data?  Is it a reasonable thing to run `dwidenoise` on a series that has multiple phase encodings in it?  At the moment, with most users just acquiring a *separate* pair of blip-up/down b=0 images, I just tell them to `dwidenoise` their main series (which has just one phase encoding direction), and leave the extra b=0 images alone; explaining them that these never interact with the main series (but are just used among/relative to themselves for estimation of the field).  If we ask users to concatenate certain parts of their data, we have to be careful about what happens with respect to this step; or be clear about the fact that things only have to be concatenated for the sake of the user interface of `dwipreproc`.  This also at least slightly begs the question on whether it's still not better to have the "`ap.mif`" and "`pa.mif`" as separate arguments (potentially to the `topup_images` option).  Users typically get them non-concatenated from their DICOM data anyway, and it's inherently an easy way of making it obvious to the scripts that they are separate phase encoding directions (and in the context of `rpe_pair`, opposing ones).  Well, just some food for thought I guess...

**Comment by thijsdhollander** [comment]
...while I'm thinking of it: to further support a name change for the `topup_images` option; it may be a good thing with a distant utopian future in mind where there are maybe alternatives on this front.  You wouldn't want to be stuck with this name then.

**Comment by Lestropie** [comment]
> I haven't gone through the whole post, but from what I see, the previous use of something along the lines of ...

Yes, I meant to provide those; they'll definitely be provided with the tag update announcement. I'm thinking we could also add to the documentation an 'interface changelist' every time we tag (since we usually withhold such changes until a tag hits). Here's old & new usages for the three:

```
dwipreproc AP dwi.mif corrected.mif -rpe_none
---
dwipreproc dwi.mif corrected.mif -rpe_none -pe_dir AP

dwipreproc AP dwi.mif corrected.mif -rpe_pair ap.mif pa.mif
---
mrcat ap.mif pa.mif pair.mif -axis 3
dwipreproc dwi.mif corrected.mif -rpe_pair -pe_dir AP -topup_images pair.mif

dwipreproc AP dwiAP.mif corrected.mif -rpe_all dwiPA.mif
---
mrcat dwiAP.mif dwiPA.mif dwi.mif -axis 3
dwipreproc dwi.mif corrected.mif -rpe_all -pe_dir AP
```

> I would change the name of the option topup_images though.

Agreed (also thought of but forgot to bring up, plus I wanted to hit send and go home). Thought about `-field_images` and `-fmap_images`, but there's a danger of being mis-interpreted as expecting a GE-based fieldmap. Something like `-rpe_images` would conflict with the existing `-rpe_*` options, and is also restrictive, since with `topup` they don't strictly have to be _reversed_ phase encoding: there just needs to be phase encoding _contrast_.

> Is topup_images actually used for anything else than in the rpe_pair context?

Yes: If using `-rpe_header`, `-topup_images` can still be used to specify images that are used for field estimation but won't form part of the output.

> Before (i.e., take my initial example above for what follows), one could supply multiple images to `-rpe_pair`, e.g. `ap.mif` and `pa.mif` could have more than one (b=0) image each, using the 4th dimension. How would this work via the `-topup_images` option, taking just one image?

If using `-rpe_header`, the phase encoding of each volume is determined from the header data and passed to `topup` accordingly. If using `-rpe_pair`, it assumes that the first _half_ of the volumes have the same phase encoding as the DWIs, and the second _half_ have the opposite encoding. So it still has the same capability, with the restriction that there needs to be an equal number of volumes in each direction - but you'd probably want this anyway, having e.g. 4 AP's and one PA might bias the field estimation.

> And maybe a more general topic to think about: what do we actually advise users to do as a denoising strategy on all these data?

My instinct is that concatenating data with different phase encoding should be OK: the spatial offsets introduced by the distortions are generally small relative to the smoothness of the noise map. But I'm not the expert here. Plus, if people concatenate their DICOM data immediately using _MRtrix3_, then the phase encoding information will be right there in the header, which leaves one less thing to be manually manipulated.

**Comment by jdtournier** [comment]
A few random thoughts, although I have to admit to feeling a bit overwhelmed with the amount of information here.. 

- agree that `-topup_images` isn't quite right. This option would be used to provide the image with different PE directions or bandwidths for estimation of the distortion field, right? How about  `dist_images` or similar, then?

- I guess if you had different PE directions than just a straight reversal, the information about this would have to come from the `-rpe_header` option? 

- while it's probably sensible to assume people will have as many PE directions up as down, there's definitely examples where that won't be the case: we often need to discard motion-corrupted b=0 images from the estimation in neonates... But I agree it's not a common use case, and besides it would probably still be possible to handle this via the `-rpe_header` option, right?

- I also think it should be OK to concatenate everything for `dwidenoise` prior to any form of processing. Where there is mismatch between the PE directions, it would simply give rise to more principle components than might have been the case otherwise, but probably still fewer overall than might have been obtained by denoising each set independently. So the overall denoising ought to perform slightly better that way - although that assertion would need to be verified...

**Comment by Lestropie** [comment]
> I have to admit to feeling a bit overwhelmed with the amount of information here.

Yeah it's a lot. But it's not my research, it's software support for how to handle this type of information, and if we get it 'wrong' the first time around we get stuck in the future trying to manage and support multiple 'standards'.

> Agree that `-topup_images` isn't quite right. This option would be used to provide the image with different PE directions or bandwidths for estimation of the distortion field, right? How about `-dist_images` or similar, then?

Something like that. Or strictly, they're EPI images for estimating the inhomogeneity field; so maybe `-epi_field_images`? That also keeps the door wider open for supporting dual-GE field image based distortion correction using a different option.

> I guess if you had different PE directions than just a straight reversal, the information about this would have to come from the `-rpe_header` option?

Yep. Or I could have option(s) that allows the user to manually provide the table; but they could equivalently run `mrconvert` with `-import_pe_table` or `-import_pe_eddy` to get the table into the image header beforehand.

> While it's probably sensible to assume people will have as many PE directions up as down, there's definitely examples where that won't be the case: we often need to discard motion-corrupted b=0 images from the estimation in neonates... But I agree it's not a common use case, and besides it would probably still be possible to handle this via the `-rpe_header` option, right?

Yes and yes. Essentially the philosophy of the script redesign is 'perform distortion correction based on any arbitrary type of phase encoding information as stored in the header; but provide additional options to allow users to manually specify the phase encoding design for only the most common schemes'. If you try to have a command-line interface for any type of phase encoding, it becomes no more user-friendly than the underlying FSL commands, the complexity of which is why I wrote the first version of this script to begin with.

P.S. If you `mrconvert` from DICOM from the outset, and subsequently remove corrupted volumes using `mrconvert`, the phase encoding table in the header will be updated accordingly.

**Comment by jdtournier** [comment]
:+1:

**Comment by Lestropie** [comment]
OK, so this thing's doing my head in a little bit.

The BIDS standard states:
"The letters “i”, “j”, “k” correspond to the first, second and third axis of the data in the NIFTI file."
OK, fine. Trouble is, *MRtrix3* internally re-jigs the image axes to obey an approximately axial RAS system, modifying image transform & permuting / flipping axes accordingly.

So... That's fine on an image open, I can read the PE scheme as-is and then perform the corresponding permutations / flips within `Header::realign_transform()` as done [here](https://github.com/MRtrix3/mrtrix3/pull/785/commits/1ecb21c2830b436b7352a993cea57a799515cbd7) to keep the PE scheme consistent with the internal image axes. But:

- If you use *MRtrix3* to export the PE table from a non-axial image, what you'll get is the permuted version that doesn't match the image on disk - unless you specify the `-dontrealign` option in `mrinfo`.

- On image write, the NIfTI handler re-jigs the transform back to make the axis strides plain and simple, but this is currently done entirely within the relevant `write()` function and doesn't affect the header instance in RAM. Which means that if you write a NIfTI image based on a header, then write the phase encoding scheme from that header to e.g. a JSON file, the two won't correspond. I could potentially detect the presence of a scheme within the `File::JSON::save()` function and re-perform the permutation... but it's clumsy, and assumes that JSON files accompany only NIfTI images, which shouldn't have to be the case.

If anybody's got an insight...

**Comment by jdtournier** [comment]
This sounds to me like it's identical to the handling of the DW scheme for NIfTI: exporting the bvecs/bvals requires special handling with knowledge of the on-disk NIfTI transform. Whatever you choose to do, it would make sense to handle these two cases in the same way, and share whatever code these two pathways have in common.

But I guess more fundamentally, the issue you're asking about is how should the JSON file be interpreted when the corresponding image is _not_ in NIfTI format. I think in this case we have to acknowledge that our in-header representation is _not_ going to be the same as the BIDS standard's, just like bvecs/bvals differ from our in-header DW scheme. IIRC, the code handling the bvecs/bvals currently treats all images the same, regardless of their on-disk format. I think that's the only sensible option here too... Otherwise what happens if you generate the JSON file for an image in NIfTI, convert that image to mif without changing anything about its axes/strides, then try to use that mif with the JSON file? If we interpret the JSON file differently depending on the image format, this use case will not work...

**Comment by Lestropie** [comment]
> IIRC, the code handling the bvecs/bvals currently treats all images the same, regardless of their on-disk format.

Depends on your meaning: Both `bvecs` / `bvals` import and export call `File::NIfTI::adjust_transform()`, so are only valid for a corresponding write to NIfTI (or similar) where the header is expected to undergo the same adjustment. Any image format export that doesn't run through `File::NIfTI::adjust_transform()` won't work with those `bvecs` / `bvals`. But whether or not the corresponding image is NIfTI is not explicitly checked.

`bvecs` / `bvals` are 'more tightly locked' to NIfTI than JSON is to NIfTI - I really wanted JSON to act as generic storage of header key/value entries. Nevertheless, I suppose for phase encoding info (at least the basic case) it's a BIDS-defined field and that's locked to NIfTI, so it should probably conform to that.

> Whatever you choose to do, it would make sense to handle these two cases in the same way, and share whatever code these two pathways have in common.

Code sharing doesn't come out cleanly enough because of the scanner-space rotation and LHS coordinate system for `bvecs`.

----------------

So what I have is:

1. A PE scheme _already present in the header_ when it goes through `realign_transform()` will be correspondingly rotated (this is necessary to get the right result for non-axial acquisition DICOM imports).

2. A JSON import assumes that the corresponding loaded NIfTI has already gone through `File::NIfTI::adjust_transform()`, and so performs the same adjustment to the PE scheme, borrowed from `DWI::load_bvecs_bvals()`.

3. A JSON export assumes that the image is about to be / has been exported to NIfTI, and so also adjusts the PE scheme according to `File::NIfTI::adjust_transform()`, but the opposite adjustment to 2. (i.e. borrowed from `DWI::save_bvecs_bvals()`).

I reckon the phase encoding data handing is going to need to come with some bright warnings before it can be trusted... At work I have 12 images acquired on the Skyra with 3 orientations, 4 phase encode directions per orientation, so I'll hopefully play with those soon. If anyone with GE / Philips data can get the phase encoding data import working it would be appreciated.

**Comment by jdtournier** [comment]
That sounds pretty much like what I had in mind. I think if we can tie the JSON export to NIfTI, that would actually be a good thing, it'll reduce the scope for all kinds of errors and confusion that would no doubt arise if we mix file types...

**Comment by Lestropie** [comment]
Well, it's 'tied' in so much as that it would be _recommended_ to only use JSON in conjunction with NIfTI. If you really wanted to lock it down, we could remove the JSON import/export command-line options from `mrconvert`, and force [read](https://github.com/MRtrix3/mrtrix3/blob/tag_0.3.16/lib/file/nifti1_utils.cpp#L217) and [write](https://github.com/MRtrix3/mrtrix3/blob/tag_0.3.16/lib/file/nifti1_utils.cpp#L421) of JSON files from within the NIfTI read/write functions; but I know you don't like having things happen automatically in the background.

**Comment by jdtournier** [comment]
Sorry, misread your earlier statement about bvecs/bvals - thought you'd said that would only _run_ if the corresponding file is NIfTI. But you clearly _didn't_ say that... and what you did say is actually how I understood it to work. So regardless, I think you current plan is fine.

**Comment by Lestropie** [comment]
This has all been merged due to #785, but let's keep an eye on things and see what happens...
